parser/src/infix.rs

```diff
@@ -0,0 +1,305 @@
+        };
+
+        loop {
+            if arg_stack.len() == 2 && op_stack.len() == 1 {
+                let lhs = arg_stack.pop().unwrap();
+                let rhs = arg_stack.pop().unwrap();
+                let op = op_stack.pop().unwrap();
+
+                arg_stack.push(make_op(lhs, op, rhs));
+            } else {
+                use std::cmp::Ordering::*;
+                use self::InfixExpr::*;
+
+                match self.next() {
+                    None => break,
+                    Some(Arg(next_expr)) => arg_stack.push(next_expr),
+                    Some(Op(next_op)) => {
+                        if op_stack.is_empty() {
+                            op_stack.push(next_op);
+                            continue;
+                        }
+
+                        let stack_op = op_stack.pop().unwrap();
+
+                        match i32::cmp(&self.get_precedence(&next_op),
+                                       &self.get_precedence(&stack_op)) {
+                            Greater => {
+                                op_stack.push(stack_op);
+                                op_stack.push(next_op);
+                            }
+                            Less => {
+                                let lhs = arg_stack.pop().unwrap();
+                                let rhs = arg_stack.pop().unwrap();
+
+                                self.next_op = Some(next_op);
+                                arg_stack.push(make_op(lhs, stack_op, rhs));
+                            }
+                            Equal => {
+                                match (self.get_fixity(&next_op), self.get_fixity(&stack_op)) {
+                                    (Fixity::Left, Fixity::Left) => {
+                                        let lhs = arg_stack.pop().unwrap();
+                                        let rhs = arg_stack.pop().unwrap();
+
+                                        self.next_op = Some(next_op);
+                                        arg_stack.push(make_op(lhs, stack_op, rhs));
+                                    }
+                                    (Fixity::Right, Fixity::Right) => {
+                                        op_stack.push(stack_op);
+                                        op_stack.push(next_op);
+                                    }
+                                    // Conflicting fixities at the same precedence level -
+                                    // this is a user error!
+                                    (Fixity::Left, Fixity::Right) |
+                                    (Fixity::Right, Fixity::Left) => panic!(),
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        assert_eq!(arg_stack.len(), 1);
+        assert_eq!(op_stack.len(), 0);
+
+        *arg_stack.pop().unwrap()
+    }
+}
+
+impl<'a, Id: 'a + Clone> Iterator for Infixes<'a, Id> {
```

Comment:
Yeah - left over from having trouble with the `Expr::Infix(lhs, op, rhs)` destructure.